### PR TITLE
Newtypes

### PR DESCRIPTION
**Pull request by robdockins**
We've had partial support for a while now for a `newtype` construct, but that support was never fully fleshed out.  This PR is for completing the job, writing associated documentation, test cases, etc.

CF #870

### CODE REVIEW

**Comment by robdockins** [comment]
@pnwamk, @david-christiansen, There are some nontrivial questions about how to represent values of newtypes in the cryptol remote API, and I'd love you have your opinions about what we should do here.

**Comment by robdockins** [comment]
@yav, @brianhuffman, aside from the questions about the remote API, I think this PR is ready to go.

**Comment by david-christiansen** [comment]
Can we have a video chat about it in the first week of January? Just so there's time to go back and forth a bit about what the feature is used for and its history so we can pick the right trade-offs.

**Comment by robdockins** [comment]
> Can we have a video chat about it in the first week of January? Just so there's time to go back and forth a bit about what the feature is used for and its history so we can pick the right trade-offs.

Sure, I'll set something up.

**Review by yav — APPROVED** [review]
I only looked at the diff on `github`, so not in a lot of detail, but it looks reasonable to me.

In the longer run we may want to represent newtypes with a single field a bit more efficiently, but let's worry about that if it every comes up.  I'd say merge it in, this way we'll get to test it by using it.

File: docs/ProgrammingCryptol/crashCourse/CrashCourse.tex [hunk 1]
```diff
@@ -3399,6 +3399,116 @@ \section{Type constraint synonyms}
...
+      r = x.real + y.real
+      i = x.imag + y.imag
+
+  cplxMul : CplxQ -> CplxQ -> CplxQ
+  cplxMul x y = CplxQ { real = r, imag = i }
+    where
+      r = x.real * y.real - x.imag * y.imag
+      i = x.real * y.imag + x.imag * y.real
+
+  cplxEq : CplxQ -> CplxQ -> CplxQ
```
Comment (robdockins) [inline docs/ProgrammingCryptol/crashCourse/CrashCourse.tex:3445]: Note to self : should be `cplxEq : CplxQ -> CplxQ -> Bit`

File: src/Cryptol/Eval/Reference.lhs [hunk 2]
```diff
@@ -506,7 +512,26 @@ the new bindings.
 >   case dDefinition d of
 >     DPrim   -> (dName d, pure (evalPrim (dName d)))
 >     DExpr e -> (dName d, evalExpr env e)
+>
+
+Newtypes
+--------
 
+At runtime, newtypes values are represented in exactly
+the same was as records.  The constructor function for
```
Comment (robdockins) [inline src/Cryptol/Eval/Reference.lhs:521]: was -> way

**Pull request merged**